svc/GridAutosizeService.js

```diff
@@ -0,0 +1,143 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {HoistService} from '@xh/hoist/core';
+import {throwIf, stripTags} from '@xh/hoist/utils/js';
+import {isFunction, isNil, isFinite, sortBy, reduce} from 'lodash';
+
+@HoistService
+export class GridAutosizeService {
+
+    _canvasContext;
+    _cellEl;
+
+    autoSizeColumns({gridModel, colIds = []}) {
+        throwIf(!gridModel, 'GridModel required for autosize');
+        if (!colIds.length) return;
+
+        const ret = [];
+        for (let i = 0; i < colIds.length; i++) {
+            const colId = colIds[i],
+                width = this.autoSizeColumn({colId, gridModel});
+
+            if (isFinite(width)) ret.push({colId, width});
+        }
+        return ret;
+    }
+
+    //------------------
+    // Implementation
+    //------------------
+    autoSizeColumn({colId, gridModel}) {
+        try {
+            const {store} = gridModel,
+                records = [...store.records, store.summaryRecord],
+                column = gridModel.findColumn(gridModel.columns, colId),
+                {autoSizeOptions, field, getValueFn, renderer, elementRenderer} = column,
+                {enabled, sampleCount, buffer, minWidth, maxWidth} = autoSizeOptions,
+                useRenderer = isFunction(renderer);
+
+            if (!enabled) return;
+
+            // Columns with element renderers are not supported
+            if (elementRenderer) return;
```

Comment:
would we consider falling back to the native agGrid resizing if someone really wanted to resize a column with an elementRenderer. (see my comment above -- might be useful for columns that have fixedWidths that result from elementRenderer)